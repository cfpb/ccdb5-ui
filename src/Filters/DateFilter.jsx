import './DateFilter.less'
import { changeDateRange } from '../actions/filter'
import CollapsibleFilter from './CollapsibleFilter'
import { connect } from 'react-redux'
import { DATE_RANGE_MIN } from '../constants'
import DateInput from '../DateInput'
import iconMap from '../iconMap'
import moment from 'moment'
import PropTypes from 'prop-types'
import React from 'react'
import { shortFormat } from '../utils'

export class DateFilter extends React.Component {
  constructor( props ) {
    super( props )

    this.state = this._validate( {
      from: props.from,
      through: props.through,
      messages: {}
    } )
  }

  componentWillReceiveProps( nextProps ) {
    const newState = {
      from: nextProps.from,
      through: nextProps.through,
      messages: {}
    }

    this.setState( this._validate( newState ) )
  }

  render() {
    return (
      <CollapsibleFilter title={ this.props.title }
                         className="aggregation date-filter">
          <div>
            <ul className="date-inputs">
            { this._renderDateInput( 'From', 'from' ) }
            { this._renderDateInput( 'Through', 'through' ) }
            </ul>
            { this._hasMessages( this.state.messages ) ?
              this._renderMessages() :
              null
            }
          </div>
        </CollapsibleFilter>
    )
  }

  // --------------------------------------------------------------------------
  // Subrender Methods

  _hasMessages( messages ) {
    return Object.keys( messages ).length > 0
  }

  _renderDateInput( label, field ) {
    const localProps = {
      'aria-describedby': 'input-error_message-' + field
    }

    return (
      <li>
          <label className="a-label a-label__heading body-copy">
            { label }
          </label>
          <DateInput min={ this.props.minimumDate }
                     max={ this.props.maximumDate }
                     onDateEntered={ this._onDateEntered.bind( this, field ) }
                     onError={ this._onError.bind( this, field ) }
                     textBoxProps={ localProps }
                     value={ this.state[field] }
          />
      </li>
    )
  }

  _renderMessages() {
    return (
      <ul className="messages">
      { Object.keys( this.state.messages ).map( field =>
          <li className="a-error-message"
              key={'input-error_message-' + field}
              id={'input-error_message-' + field}
              role="alert">
            <span aria-hidden="true">
              { iconMap.getIcon( 'delete-round', 'cf-icon-delete-round' ) }
            </span>

            { this.state.messages[field] }
          </li>
         ) }
      </ul>
    )
  }

  // --------------------------------------------------------------------------
  // Validation methods

  _validate( state ) {

    // Check for range errors
    const from = moment( state.from, 'MM-DD-YYYY' )
    const through = moment( state.through, 'MM-DD-YYYY' )
    if ( from && through && from > through ) {
      state.messages.ordered = "'From' must be less than 'Through'"
    } else {
      delete state.messages.ordered
    }

    return state
  }

  // --------------------------------------------------------------------------
  // DateInput interface methods

  _onDateEntered( field, date ) {
    let state = {
      from: this.state.from,
      through: this.state.through,
      messages: { ...this.state.messages }
    }

    // Update the correct field
    state[field] = shortFormat( date )

    // Clear any messages for that field
    delete state.messages[field]

    state = this._validate( state )

    this.setState( state )

    // If it's good, send an update
    if ( this._hasMessages( state.messages ) === false ) {
      const from = moment( state.from, 'MM-DD-YYYY' )
      const through = moment( state.through, 'MM-DD-YYYY' )
      const dateFrom = from.isValid() ? from.toDate() : null
      const dateThrough = through.isValid() ? through.toDate() : null
      this.props.changeDateRange( dateFrom, dateThrough )
    }
  }

  _onError( field, error, value ) {
    const messages = { ...this.state.messages }
    messages[field] = error
    this.setState( { messages, [field]: value } )
  }
}

// ----------------------------------------------------------------------------
// Meta

DateFilter.propTypes = {
  fieldName: PropTypes.string.isRequired,
  from: PropTypes.string,
  maximumDate: PropTypes.instanceOf( Date ),
  minimumDate: PropTypes.instanceOf( Date ),
  through: PropTypes.string,
  title: PropTypes.string.isRequired
}

DateFilter.defaultProps = {
  from: '',
  maximumDate: null,
  minimumDate: new Date( DATE_RANGE_MIN ),
  through: ''
}

export const mapStateToProps = ( state, ownProps ) => ( {
  from: shortFormat( state.query[ownProps.fieldName + '_min'] ),
  through: shortFormat( state.query[ownProps.fieldName + '_max'] )
} )

export const mapDispatchToProps = ( dispatch, ownProps ) => ( {
  changeDateRange: ( from, through ) => {
    dispatch( changeDateRange( ownProps.fieldName, from, through ) )
  }
} )

export default connect( mapStateToProps, mapDispatchToProps )( DateFilter )
