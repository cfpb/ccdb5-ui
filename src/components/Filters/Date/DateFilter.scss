@use 'css/base' as *;

.date-filter {
  h4 {
    margin-bottom: 0;
  }

  ul.date-inputs {
    margin: 0;

    li {
      display: inline-block;
      width: calc(50% - 8px); // Half $gutter-normal

      list-style: none;

      &:first-of-type {
        margin-right: $gutter-normal;
      }
      input {
        width: 100%;
      }
    }
  }

  ul.messages {
    padding-top: $gutter-narrow;

    li {
      margin: 0;

      list-style: none;
    }
  }
}
