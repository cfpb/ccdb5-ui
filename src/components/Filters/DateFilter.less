@import (less) '../../css/base.less';

.date-filter {
  h4 {
    margin-bottom:0;
  }

  ul.date-inputs {
    margin: 0;

    li {
      display: inline-block;
      width: ~'calc(50% - 8px)'; // Half @gutter-normal

      list-style: none;

      &:first-of-type {
        margin-right: @gutter-normal;
      }

      .m-btn-inside-input {
        .a-text-input {
          padding-right: @grid_gutter-width;
        }
        .a-btn {
          color: @gray;
        }
      }
    }
  }

  ul.messages {
    padding-top: @gutter-narrow;

    li {
      margin: 0;

      list-style: none;
    }

    .a-error-message {
      .cf-icon-svg {
        padding-right: @gutter-minimum;
      }
    }
  }
}
