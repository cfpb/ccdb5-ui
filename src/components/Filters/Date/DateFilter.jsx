/* eslint complexity: ["error", 8] */
import './DateFilter.scss';
import { DATE_VALIDATION_FORMAT, maxDate, minDate } from '../../../constants';
import { useEffect, useMemo, useRef, useState } from 'react';
import {
  selectQueryDateReceivedMax,
  selectQueryDateReceivedMin,
} from '../../../reducers/query/selectors';
import { useDispatch, useSelector } from 'react-redux';
import { datesChanged } from '../../../reducers/query/querySlice';
import { CollapsibleFilter } from '../CollapsibleFilter/CollapsibleFilter';
import { DateRanges } from './DateRanges';
import dayjs from 'dayjs';
import dayjsCustomParseFormat from 'dayjs/plugin/customParseFormat';
import dayjsIsBetween from 'dayjs/plugin/isBetween';
import dayjsUtc from 'dayjs/plugin/utc';
import { formatDateModel } from '../../../utils/formatDate';
import getIcon from '../../Common/Icon/iconMap';

dayjs.extend(dayjsCustomParseFormat);
dayjs.extend(dayjsIsBetween);
dayjs.extend(dayjsUtc);

const WARN_SERIES_BREAK =
  'CFPB updated product and issue options in April 2017 and August 2023.';

const LEARN_SERIES_BREAK =
  'https://www.consumerfinance.gov/data-research/consumer-complaints/#past-changes';

export const DateFilter = () => {
  const fieldName = 'date_received';
  const title = 'Date CFPB received the complaint';
  const dateFrom = useSelector(selectQueryDateReceivedMin);
  const dateThrough = useSelector(selectQueryDateReceivedMax);
  const initialFromDate = dayjs(dateFrom).isValid()
    ? formatDateModel(dateFrom)
    : '';
  const initialThroughDate = dayjs(dateThrough).isValid()
    ? formatDateModel(dateThrough)
    : '';
  const [fromDate, setFromDate] = useState(initialFromDate);
  const [throughDate, setThroughDate] = useState(initialThroughDate);
  const dispatch = useDispatch();

  const errorMessageText = "'From' date must be less than 'through' date";
  const errorSameDate = "'From' date cannot be the same as 'Through' date";

  const fromRef = useRef();
  const throughRef = useRef();

  useEffect(() => {
    // put it in YYYY-MM-DD format
    // validate to make sure it's not invalid
    const validFromDate = dateFrom ? formatDateModel(dateFrom) : '';
    setFromDate(validFromDate);
  }, [dateFrom]);

  useEffect(() => {
    const validThroughDate = dateThrough ? formatDateModel(dateThrough) : '';
    setThroughDate(validThroughDate);
  }, [dateThrough]);

  const handleKeyDownFromDate = (event) => {
    if (event.key === 'Enter') {
      fromRef.current.blur();
    }
  };

  const handleKeyDownThroughDate = (event) => {
    if (event.key === 'Enter') {
      throughRef.current.blur();
    }
  };

  const errors = useMemo(() => {
    if (dayjs(fromDate).isAfter(throughDate)) {
      return errorMessageText;
    }
    if (dayjs(fromDate).isSame(throughDate)) {
      return errorSameDate;
    }
    return false;
  }, [fromDate, throughDate]);

  const handleDateChange = () => {
    // setFromDate and setThroughDate do not update the state quick enough
    // to be used here
    let _fromDate = fromDate;
    let _throughDate = throughDate;
    // don't do anything if its empty
    if (_fromDate < minDate && _fromDate) {
      fromRef.current.value = minDate;
      _fromDate = minDate;
    }
    if (_throughDate > maxDate && _throughDate) {
      throughRef.current.value = maxDate;
      _throughDate = maxDate;
    }

    const isDateDifferent =
      dateFrom !== _fromDate || dateThrough !== _throughDate;
    if (dayjs(_throughDate).isAfter(_fromDate) && isDateDifferent) {
      dispatch(datesChanged(_fromDate, _throughDate));
    }
  };

  const inputFromClassName = useMemo(() => {
    const style = ['a-text-input'];
    if (
      dayjs(fromDate).isBefore(minDate) ||
      dayjs(fromDate).isAfter(throughDate) ||
      dayjs(fromDate).isSame(throughDate)
    ) {
      style.push('a-text-input--error');
    }
    return style.join(' ');
  }, [fromDate, throughDate]);

  const inputThroughClassName = useMemo(() => {
    const style = ['a-text-input'];
    if (
      dayjs(throughDate).isAfter(maxDate) ||
      dayjs(throughDate).isBefore(fromDate) ||
      dayjs(throughDate).isSame(fromDate)
    ) {
      style.push('a-text-input--error');
    }
    return style.join(' ');
  }, [fromDate, throughDate]);

  return (
    <CollapsibleFilter
      title={title}
      className="aggregation date-filter"
      desc=""
    >
      <div>
        <p className="u-mt15">
          {' '}
          {WARN_SERIES_BREAK}{' '}
          <a
            href={LEARN_SERIES_BREAK}
            target="_blank"
            rel="noopener noreferrer"
            aria-label="Learn more about Product and
                  Issue changes (opens in new window)"
          >
            Learn More
          </a>
        </p>
        <ul className="date-inputs">
          <li>
            <label
              className="a-label a-label__heading body-copy"
              htmlFor={`${fieldName}-from`}
            >
              From
            </label>
            <div className="o-search-input">
              <div className="o-search-input__input">
                <input
                  id={`${fieldName}-from`}
                  className={inputFromClassName}
                  onBlur={handleDateChange}
                  onChange={(evt) => {
                    const value = evt.target.value || minDate;
                    setFromDate(value);
                  }}
                  onKeyDown={handleKeyDownFromDate}
                  min={minDate}
                  max={maxDate}
                  ref={fromRef}
                  placeholder={DATE_VALIDATION_FORMAT}
                  type="date"
                  value={fromDate}
                />
              </div>
            </div>
          </li>
          <li>
            <label
              className="a-label a-label__heading body-copy"
              htmlFor={`${fieldName}-through`}
            >
              Through
            </label>
            <div className="o-search-input">
              <div className="o-search-input__input">
                <input
                  id={`${fieldName}-through`}
                  className={inputThroughClassName}
                  onBlur={handleDateChange}
                  onChange={(evt) => {
                    const value = evt.target.value || minDate;
                    setThroughDate(value);
                  }}
                  onKeyDown={handleKeyDownThroughDate}
                  min={minDate}
                  max={maxDate}
                  placeholder={DATE_VALIDATION_FORMAT}
                  ref={throughRef}
                  type="date"
                  value={throughDate}
                />
              </div>
            </div>
          </li>
        </ul>
        <DateRanges />
        {errors ? (
          <div className="a-form-alert a-form-alert--error" role="alert">
            <span aria-hidden="true">
              {getIcon('delete-round', 'cf-icon-delete-round')}
            </span>
            <span className="a-form-alert__text">{errors + ' '}</span>
          </div>
        ) : null}
      </div>
    </CollapsibleFilter>
  );
};
