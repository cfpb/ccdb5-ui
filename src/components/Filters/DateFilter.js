/* eslint complexity: ["error", 8] */
import './DateFilter.less';
import { DATE_VALIDATION_FORMAT, maxDate, minDate } from '../../constants';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import {
  selectQueryDateReceivedMax,
  selectQueryDateReceivedMin,
} from '../../reducers/query/selectors';
import { useDispatch, useSelector } from 'react-redux';
import { changeDates } from '../../actions/filter';
import CollapsibleFilter from './CollapsibleFilter';
import { DateRanges } from './DateRanges';
import dayjs from 'dayjs';
import dayjsCustomParseFormat from 'dayjs/plugin/customParseFormat';
import dayjsIsBetween from 'dayjs/plugin/isBetween';
import dayjsUtc from 'dayjs/plugin/utc';
import { formatDate } from '../../utils/formatDate';
import iconMap from '../iconMap';

dayjs.extend(dayjsCustomParseFormat);
dayjs.extend(dayjsIsBetween);
dayjs.extend(dayjsUtc);

const WARN_SERIES_BREAK =
  'CFPB updated product and issue options in April 2017 and August 2023.';

const LEARN_SERIES_BREAK =
  'https://www.consumerfinance.gov/data-research/consumer-complaints/#past-changes';

export const DateFilter = () => {
  const fieldName = 'date_received';
  const title = 'Date CFPB received the complaint';
  const dateFrom = useSelector(selectQueryDateReceivedMin);
  const dateThrough = useSelector(selectQueryDateReceivedMax);
  const initialFromDate = dayjs(dateFrom).isValid() ? formatDate(dateFrom) : '';
  const initialThroughDate = dayjs(dateThrough).isValid()
    ? formatDate(dateThrough)
    : '';
  const [fromDate, setFromDate] = useState(initialFromDate);
  const [throughDate, setThroughDate] = useState(initialThroughDate);
  const dispatch = useDispatch();

  const errorMessageText = "'From' date must be less than 'through' date";
  const errorSameDate = "'From' date cannot be the same as 'Through' date";

  const fromRef = useRef();
  const throughRef = useRef();

  useEffect(() => {
    // put it in YYYY-MM-DD format
    // validate to make sure it's not invalid
    const validFromDate = dateFrom ? formatDate(dateFrom) : '';
    setFromDate(validFromDate);
  }, [dateFrom]);

  useEffect(() => {
    const validThroughDate = dateThrough ? formatDate(dateThrough) : '';
    setThroughDate(validThroughDate);
  }, [dateThrough]);

  const handleClear = (period) => {
    if (period === 'from') {
      dispatch(changeDates(fieldName, minDate, throughDate));
    }
    if (period === 'through') {
      dispatch(changeDates(fieldName, fromDate, maxDate));
    }
  };

  const handleKeyDownFromDate = (event) => {
    if (event.key === 'Enter') {
      fromRef.current.blur();
    }
  };

  const handleKeyDownThroughDate = (event) => {
    if (event.key === 'Enter') {
      throughRef.current.blur();
    }
  };

  const errors = useMemo(() => {
    if (dayjs(fromDate).isAfter(throughDate)) {
      return errorMessageText;
    }
    if (dayjs(fromDate).isSame(throughDate)) {
      return errorSameDate;
    }
    return false;
  }, [fromDate, throughDate]);

  const handleDateChange = () => {
    // setFromDate and setThroughDate do not update the state quick enough
    // to be used here
    let _fromDate = fromDate;
    let _throughDate = throughDate;
    // don't do anything if its empty
    if (_fromDate < minDate && _fromDate) {
      fromRef.current.value = minDate;
      _fromDate = minDate;
    }
    if (_throughDate > maxDate && _throughDate) {
      throughRef.current.value = maxDate;
      _throughDate = maxDate;
    }

    const isDateDifferent =
      dateFrom !== _fromDate || dateThrough !== _throughDate;
    if (dayjs(_throughDate).isAfter(_fromDate) && isDateDifferent) {
      dispatch(changeDates(fieldName, _fromDate, _throughDate));
    }
  };

  const inputFromClassName = useMemo(() => {
    const style = ['a-text-input'];
    if (
      dayjs(fromDate).isBefore(minDate) ||
      dayjs(fromDate).isAfter(throughDate) ||
      dayjs(fromDate).isSame(throughDate)
    ) {
      style.push('a-text-input__error');
    }
    return style.join(' ');
  }, [fromDate, throughDate]);

  const inputThroughClassName = useMemo(() => {
    const style = ['a-text-input'];
    if (
      dayjs(throughDate).isAfter(maxDate) ||
      dayjs(throughDate).isBefore(fromDate) ||
      dayjs(throughDate).isSame(fromDate)
    ) {
      style.push('a-text-input__error');
    }
    return style.join(' ');
  }, [fromDate, throughDate]);

  return (
    <CollapsibleFilter title={title} className="aggregation date-filter">
      <div>
        <p className="u-mt15">
          {' '}
          {WARN_SERIES_BREAK}{' '}
          <a
            href={LEARN_SERIES_BREAK}
            target="_blank"
            rel="noopener noreferrer"
            aria-label="Learn more about Product and
                  Issue changes (opens in new window)"
          >
            Learn More
          </a>
        </p>
        <ul className="date-inputs">
          <li>
            <label
              className="a-label a-label__heading body-copy"
              htmlFor={`${fieldName}-from`}
            >
              From
            </label>
            <div className="m-btn-inside-input">
              <input
                id={`${fieldName}-from`}
                className={inputFromClassName}
                onBlur={handleDateChange}
                onChange={(evt) => setFromDate(evt.target.value)}
                onKeyDown={handleKeyDownFromDate}
                min={minDate}
                max={maxDate}
                ref={fromRef}
                placeholder={DATE_VALIDATION_FORMAT}
                type="date"
                value={fromDate}
              />
              <button
                className="a-btn a-btn__link"
                onClick={() => handleClear('from')}
              >
                {iconMap.getIcon('delete')}
                <span className="u-visually-hidden">
                  Clear date received from filter
                </span>
              </button>
            </div>
          </li>
          <li>
            <label
              className="a-label a-label__heading body-copy"
              htmlFor={`${fieldName}-through`}
            >
              Through
            </label>
            <div className="m-btn-inside-input">
              <input
                id={`${fieldName}-through`}
                className={inputThroughClassName}
                onBlur={handleDateChange}
                onChange={(evt) => setThroughDate(evt.target.value)}
                onKeyDown={handleKeyDownThroughDate}
                min={minDate}
                max={maxDate}
                placeholder={DATE_VALIDATION_FORMAT}
                ref={throughRef}
                type="date"
                value={throughDate}
              />
              <button
                className="a-btn a-btn__link"
                onClick={() => handleClear('through')}
              >
                {iconMap.getIcon('delete')}
                <span className="u-visually-hidden">
                  Clear date received through filter
                </span>
              </button>
            </div>
          </li>
        </ul>
        <DateRanges />
        {errors ? (
          <>
            {errors + ' '}
            <span aria-hidden="true">
              {iconMap.getIcon('delete-round', 'cf-icon-delete-round')}
            </span>
          </>
        ) : null}
      </div>
    </CollapsibleFilter>
  );
};
