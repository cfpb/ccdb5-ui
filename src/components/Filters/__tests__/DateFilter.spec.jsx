import { DateFilter } from '../DateFilter';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import React from 'react';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';

function setupSnapshot(query) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    query: query,
  });

  return renderer.create(
    <Provider store={store}>
      <DateFilter
        fieldName="date_received"
        title="Date CFPB Received the complaint"
      />
    </Provider>
  );
}

describe('component::DateFilter', () => {
  describe('snapshot', () => {
    it('supports no dates in the Redux Store', () => {
      const target = setupSnapshot({});
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('shows errors', () => {
      const target = setupSnapshot({
        date_received_min: new Date(2016, 0, 1),
        date_received_max: new Date(2000, 0, 1),
      });
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('shows a warning for April 2017', () => {
      const target = setupSnapshot({
        date_received_min: new Date(2016, 0, 1),
        date_received_max: new Date(2018, 0, 1),
      });
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
