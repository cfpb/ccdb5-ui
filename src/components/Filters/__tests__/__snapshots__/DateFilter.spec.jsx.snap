// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::DateFilter snapshot shows a warning for April 2017 1`] = `
<section
  className="o-expandable aggregation date-filter"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Date CFPB Received the complaint
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        aria-label="Hide Date CFPB Received the complaint filter"
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p />
  <div>
    <ul
      className="date-inputs"
    >
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-from"
        >
          From
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-from"
            max={2019-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value="1/1/2016"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <svg
              className="cf-icon-svg"
              viewBox="0 0 718.9 1200"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
              />
            </svg>
            <span
              className="u-visually-hidden"
            >
              Clear
            </span>
          </button>
        </div>
      </li>
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-through"
        >
          Through
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-through"
            max={2019-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value="1/1/2018"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <svg
              className="cf-icon-svg"
              viewBox="0 0 718.9 1200"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
              />
            </svg>
            <span
              className="u-visually-hidden"
            >
              Clear
            </span>
          </button>
        </div>
      </li>
    </ul>
    <p>
       
      CFPB updated product and issue options available to consumers in April 2017 
      <a
        aria-label="Learn more about Product and Issue changes (opens in new window)"
        href="https://files.consumerfinance.gov/f/documents/201704_cfpb_Summary_of_Product_and_Sub-product_Changes.pdf"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn More
      </a>
    </p>
  </div>
</section>
`;

exports[`component::DateFilter snapshot shows errors 1`] = `
<section
  className="o-expandable aggregation date-filter"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Date CFPB Received the complaint
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        aria-label="Hide Date CFPB Received the complaint filter"
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p />
  <div>
    <ul
      className="date-inputs"
    >
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-from"
        >
          From
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-from"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value="1/1/2016"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <svg
              className="cf-icon-svg"
              viewBox="0 0 718.9 1200"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
              />
            </svg>
            <span
              className="u-visually-hidden"
            >
              Clear
            </span>
          </button>
        </div>
      </li>
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-through"
        >
          Through
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-through"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value="1/1/2000"
          />
          <button
            className="a-btn a-btn__link"
            onClick={[Function]}
          >
            <svg
              className="cf-icon-svg"
              viewBox="0 0 718.9 1200"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M451.4 613.7l248.1-248.1c25.6-25.1 26-66.3.8-91.9s-66.3-26-91.9-.8l-.8.8-248.1 248.1-248.1-248.1c-25.4-25.4-66.5-25.4-91.9 0s-25.4 66.5 0 91.9l248.1 248.1L19.5 861.8c-25.6 25.1-26 66.3-.8 91.9s66.3 26 91.9.8l.8-.8 248.1-248.1 248.1 248.1c25.4 25.4 66.5 25.4 91.9 0s25.4-66.5 0-91.9L451.4 613.7z"
              />
            </svg>
            <span
              className="u-visually-hidden"
            >
              Clear
            </span>
          </button>
        </div>
      </li>
    </ul>
    <ul
      className="messages"
    >
      <li
        className="a-error-message"
        id="input-error_message-ordered"
        role="alert"
      >
        <span
          aria-hidden="true"
        >
          <svg
            className="cf-icon-svg cf-icon-delete-round"
            viewBox="0 0 1000 1200"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm261.8 692.2c19.4 19.6 19.3 51.3-.3 70.7-19.5 19.3-50.9 19.3-70.4 0L499.6 676.6 308 868.1c-19.6 19.4-51.3 19.3-70.7-.3-19.3-19.5-19.3-50.9 0-70.4l191.6-191.5-191.6-191.6c-19.3-19.8-18.9-51.4.9-70.7 19.4-18.9 50.4-18.9 69.8 0l191.6 191.5 191.5-191.5c19.6-19.4 51.3-19.3 70.7.3 19.3 19.5 19.3 50.9 0 70.4L570.3 605.9l191.5 191.5z"
            />
          </svg>
        </span>
        'From' must be less than 'Through'
      </li>
      <li
        className="a-error-message"
        id="input-error_message-through"
        role="alert"
      >
        <span
          aria-hidden="true"
        >
          <svg
            className="cf-icon-svg cf-icon-delete-round"
            viewBox="0 0 1000 1200"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm261.8 692.2c19.4 19.6 19.3 51.3-.3 70.7-19.5 19.3-50.9 19.3-70.4 0L499.6 676.6 308 868.1c-19.6 19.4-51.3 19.3-70.7-.3-19.3-19.5-19.3-50.9 0-70.4l191.6-191.5-191.6-191.6c-19.3-19.8-18.9-51.4.9-70.7 19.4-18.9 50.4-18.9 69.8 0l191.6 191.5 191.5-191.5c19.6-19.4 51.3-19.3 70.7.3 19.3 19.5 19.3 50.9 0 70.4L570.3 605.9l191.5 191.5z"
            />
          </svg>
        </span>
        '1/1/2000' must be greater than 1/1/2015
      </li>
    </ul>
  </div>
</section>
`;

exports[`component::DateFilter snapshot supports no dates in the Redux Store 1`] = `
<section
  className="o-expandable aggregation date-filter"
>
  <div
    className="o-expandable_header o-expandable_target"
  >
    <h4
      className="o-expandable_header-left o-expandable_label"
    >
      Date CFPB Received the complaint
    </h4>
    <span
      className="o-expandable_header-right o-expandable_link"
    >
      <button
        aria-label="Hide Date CFPB Received the complaint filter"
        className="a-btn a-btn__link o-expandable_cue o-expandable_cue-close"
        onClick={[Function]}
      >
        Hide
        <svg
          className="cf-icon-svg"
          viewBox="0 0 1000 1200"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500 105.2c-276.1 0-500 223.9-500 500s223.9 500 500 500 500-223.9 500-500-223.9-500-500-500zm263.1 550.7H236c-27.6 0-50-22.4-50-50s22.4-50 50-50h527.1c27.6 0 50 22.4 50 50s-22.4 50-50 50z"
          />
        </svg>
      </button>
    </span>
  </div>
  <p />
  <div>
    <ul
      className="date-inputs"
    >
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-from"
        >
          From
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-from"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value=""
          />
        </div>
      </li>
      <li>
        <label
          className="a-label a-label__heading body-copy"
          htmlFor="date_received-through"
        >
          Through
        </label>
        <div
          className="m-btn-inside-input"
        >
          <input
            className="a-text-input"
            id="date_received-through"
            max={2015-12-31T12:00:00.000Z}
            min={2015-01-01T12:00:00.000Z}
            onChange={[Function]}
            placeholder="M/D/YYYY"
            size="15"
            value=""
          />
        </div>
      </li>
    </ul>
  </div>
</section>
`;
